test/e2e/upgrade_test.go

```diff
@@ -0,0 +1,131 @@
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"testing"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	nc "github.com/openshift/windows-machine-config-operator/pkg/controller/windowsmachine/nodeconfig"
+	framework "github.com/operator-framework/operator-sdk/pkg/test"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"log"
+)
+
+func upgradeTestSuite(t *testing.T) {
+
+	t.Run("Node deletion during upgrade", func(t *testing.T) { testUpgradeVersion(t) })
+	t.Run("Node deletion during annotation tampering", func(t *testing.T) { testTamperAnnotation(t) })
+}
+
+// testUpgradeVersion tests if the operator deletes machines and recreats them, if the node annotation doesn`t match
+// with the expected annotation
+func testUpgradeVersion(t *testing.T) {
+	testCtx, err := NewTestContext(t)
+	require.NoError(t, err)
+
+	operatorVersion, err := testCtx.getWMCOVersion()
+	require.NoError(t, err, "could not get WMCO version")
+
+	err = testCtx.waitForWindowsNodesAnnotation(gc.numberOfNodes, operatorVersion)
+	if err != nil {
+		t.Fatalf("windows node upgrade failed  with %v", err)
+	}
+
+	// Get all the Windows nodes
+	nodes, err := framework.Global.KubeClient.CoreV1().Nodes().List(context.TODO(),
+		metav1.ListOptions{LabelSelector: nc.WindowsOSLabel})
+	require.NoError(t, err)
+
+	for _, node := range nodes.Items {
+		t.Run(node.GetName(), func(t *testing.T) {
+			require.Containsf(t, node.Annotations, nc.VersionAnnotation, "node %s missing version annotation", node.GetName())
+			assert.Equal(t, operatorVersion, node.Annotations[nc.VersionAnnotation], "Node recreation failed. WMCO version annotation mismatch")
+		})
+	}
+}
+
+// testTamperAnnotation tests if the operator deletes machines and recreats them, if the node annotation is changed to an invalid value
+// with the expected annotation
+func testTamperAnnotation(t *testing.T) {
+	testCtx, err := NewTestContext(t)
+	require.NoError(t, err)
+
+	operatorVersion, err := testCtx.getWMCOVersion()
+	require.NoError(t, err, "could not get WMCO version")
+
+	// tamper node annotation
+	nodes, err := framework.Global.KubeClient.CoreV1().Nodes().List(context.TODO(),
+		metav1.ListOptions{LabelSelector: nc.WindowsOSLabel})
+	require.NoError(t, err)
+
+	for _, node := range nodes.Items {
+		patchData := fmt.Sprintf(`{"metadata":{"annotations":{"%s":"%s"}}}`, nc.VersionAnnotation, "0.0.2+baa8c1f-dirty")
```

Comment:
There is potential for `0.0.2+baa8c1f-dirty` to be an actual version :-) I suggest something completely bogus like "bad-version"